cli: better errors

tbg opened the pull request:
Removed `panicGuard` and `mustUsage` and introduced proper error handling,
cleaning up some cruft along the way. However, the primary goal was exposing
better errors to clients. To this end, GRPC errors are now special-cased
and print a legible error. For example,

```
$ ./cockroach node ls # node is not even running
Error: unable to connect or connection lost.

Please check the address and credentials such as certificates (if attempting to
communicate with a secure cluster).

(error code 14: grpc: the connection is unavailable)
Failed running "node"
```

In particular, I removed the `failfast=false` toggle in RPCSender, which (at
least for the cli debug commands, where it is used) would have calls hang
forever, even if the server wasn't even running or the certificates not
specified.

Updated `grpcutil.IsConnectionClosed()` to operate on the `errors.Cause` of
an error.

Reworked `runQuit`:

```
./cockroach quit --host gamma.gce.cockroachdb.com # host alive, but secure
Error: unable to connect or connection lost.

Please check the address and credentials such as certificates (if attempting to
communicate with a secure cluster).

(error code 13: transport is closing)
Failed running "quit"

$ ./cockroach quit # host not running
Error: unable to connect or connection lost.

Please check the address and credentials such as certificates (if attempting to
communicate with a secure cluster).

(error code 14: grpc: the connection is unavailable)
Failed running "quit"

$  $ ./cockroach quit # node running
ok
```

Fixes #9151.
Fixes #9153.

<!-- Reviewable:start -->

---

This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/9917)

<!-- Reviewable:end -->

tamird commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> hurray!

---

Reviewed 20 of 20 files at r1.
Review status: 14 of 20 files reviewed at latest revision, 4 unresolved discussions, some commit checks pending.

---

*[cli/kv.go, line 224 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/9917#-KTtu-ext5cKkhVm4bEM:-KTtu-eySHrxqacJKi-6:btxvznn) ([raw file](https://github.com/cockroachdb/cockroach/blob/053ce1bc478cdd6b373d8fce1e2709ce54d32aa1/cli/kv.go#L224)):*
> ```Go
> 		var err error
> 		if amount, err = strconv.Atoi(args[1]); err != nil {
> 			return errors.Errorf(
> ```

not Wrapf? i liked the old error message better, the new one reads like "b" is the key which does not contain an integer.

---

*[cli/kv.go, line 237 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/9917#-KTtuBgdy_ANyIoM3G1w:-KTtuBgdy_ANyIoM3G1x:b-riqpdi) ([raw file](https://github.com/cockroachdb/cockroach/blob/053ce1bc478cdd6b373d8fce1e2709ce54d32aa1/cli/kv.go#L237)):*
> ```Go
> 	r, err := kvDB.Inc(context.TODO(), key, int64(amount))
> 	if err != nil {
> 		return errors.Wrap(err, "increment failed")
> ```

is this tested? i'm wondering what kind of error it produces.

---

*[cli/start.go, line 541 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/9917#-KTtuibahgxSthtxAlDB:-KTtuibahgxSthtxAlDC:bqszx9e) ([raw file](https://github.com/cockroachdb/cockroach/blob/053ce1bc478cdd6b373d8fce1e2709ce54d32aa1/cli/start.go#L541)):*
> ```Go
> 	// server (in which case we don't want our caller to try to come back
> 	// with a hard retry) and the case in which an attempt to shut down
> 	// fails (times out, or perhaps
> ```

finish the thought

---

*[server/admin.go, line 889 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/9917#-KTttOAfXMeYM4UP00-S:-KTttOAfXMeYM4UP00-T:bjbqup9) ([raw file](https://github.com/cockroachdb/cockroach/blob/053ce1bc478cdd6b373d8fce1e2709ce54d32aa1/server/admin.go#L889)):*
> ```Go
> 		go s.server.stopper.Stop()
> 	} else {
> 		return nil
> ```

how come? the old behaviour was to wait for the drain; don't we still want that?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/9917#-:-KTtvLfpC-DKJ64tFzPW:b-7sqz7)*
<!-- Sent from Reviewable.io -->

tamird commented:
Reviewed 6 of 6 files at r2.
Review status: all files reviewed at latest revision, 4 unresolved discussions, some commit checks pending.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/9917)*
<!-- Sent from Reviewable.io -->

a-robinson commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Reviewed 14 of 20 files at r1, 6 of 6 files at r2.
Review status: all files reviewed at latest revision, 5 unresolved discussions, all commit checks successful.

---

*[cli/range.go, line 75 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/9917#-KTtyka9d_6f9vVJS_00:-KTtyka9d_6f9vVJS_01:bi1fzc4) ([raw file](https://github.com/cockroachdb/cockroach/blob/11e5d4f7cb8dec34a582a733c8d3d56bc78ea705/cli/range.go#L75)):*
> ```Go
> 		desc := &roachpb.RangeDescriptor{}
> 		if err := row.ValueProto(desc); err != nil {
> 			return errors.Wrapf(err, "unable to unmarshal range descriptor at %s\n", row.Key)
> ```

Probably don't need the newline anymore

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/9917#-:-KTtzpWSI4uYAB9sFX8D:bnfp4nl)*
<!-- Sent from Reviewable.io -->

tschottdorf commented:
Review status: all files reviewed at latest revision, 5 unresolved discussions, all commit checks successful.

---

*[cli/kv.go, line 224 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/9917#-KTtu-ext5cKkhVm4bEM:-KTty7DxhSCli3UhthHV:b-lfpbmw) ([raw file](https://github.com/cockroachdb/cockroach/blob/053ce1bc478cdd6b373d8fce1e2709ce54d32aa1/cli/kv.go#L224)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> not Wrapf? i liked the old error message better, the new one reads like "b" is the key which does not contain an integer.
</details>
Changed to:

```
$ ./cockroach debug kv inc banana asd
Error: invalid increment: strconv.ParseInt: parsing "asd": invalid syntax
```

---

*[cli/kv.go, line 237 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/9917#-KTtuBgdy_ANyIoM3G1w:-KTtx5wyFrG3z-pFz0KI:b-mtdh8u) ([raw file](https://github.com/cockroachdb/cockroach/blob/053ce1bc478cdd6b373d8fce1e2709ce54d32aa1/cli/kv.go#L237)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> is this tested? i'm wondering what kind of error it produces.
</details>
```
./cockroach debug kv inc a
Error: roachpb.Batch RPC failed: rpc error: code = 14 desc = grpc: the connection is unavailable
Failed running "debug"
```

and, of course, `err.Wrap(nil, anything) == nil`.

---

*[cli/range.go, line 75 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/9917#-KTtyka9d_6f9vVJS_00:-KTu-omyVm8NnRr2MKfQ:bss4a2b) ([raw file](https://github.com/cockroachdb/cockroach/blob/11e5d4f7cb8dec34a582a733c8d3d56bc78ea705/cli/range.go#L75)):*
<details><summary><i>Previously, a-robinson (Alex Robinson) wrote…</i></summary>
> Probably don't need the newline anymore
</details>
Removed.

---

*[cli/start.go, line 541 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/9917#-KTtuibahgxSthtxAlDB:-KTtxaADn-goMHai9BaO:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/053ce1bc478cdd6b373d8fce1e2709ce54d32aa1/cli/start.go#L541)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> finish the thought
</details>
Done.

---

*[server/admin.go, line 889 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/9917#-KTttOAfXMeYM4UP00-S:-KTtxg-vKScR6Qz04YVA:b-7zza00) ([raw file](https://github.com/cockroachdb/cockroach/blob/053ce1bc478cdd6b373d8fce1e2709ce54d32aa1/server/admin.go#L889)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> how come? the old behaviour was to wait for the drain; don't we still want that?
</details>
Which drain? The only remaining code below waits for the stopper to stop, which isn't going to happen unless `req.Shutdown` is true. We just weren't ever calling this with `Shutdown: false` before, so nobody noticed.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/9917)*
<!-- Sent from Reviewable.io -->

tamird commented:
Reviewed 5 of 5 files at r3.
Review status: all files reviewed at latest revision, 3 unresolved discussions, some commit checks pending.

---

*[cli/kv.go, line 237 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/9917#-KTtuBgdy_ANyIoM3G1w:-KTu0ByMXx-kfGG5Ske5:bplseij) ([raw file](https://github.com/cockroachdb/cockroach/blob/053ce1bc478cdd6b373d8fce1e2709ce54d32aa1/cli/kv.go#L237)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary>
> ```
> ./cockroach debug kv inc a
> Error: roachpb.Batch RPC failed: rpc error: code = 14 desc = grpc: the connection is unavailable
> Failed running "debug"
> ```
> 
> and, of course, `err.Wrap(nil, anything) == nil`.
</details>
Wait, that's the error you got? Seems dubious since it doesn't contain `increment failed`.

---

*[server/admin.go, line 889 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/9917#-KTttOAfXMeYM4UP00-S:-KTu0fxn5IU_d1_NfV6T:b-uyp5h6) ([raw file](https://github.com/cockroachdb/cockroach/blob/053ce1bc478cdd6b373d8fce1e2709ce54d32aa1/server/admin.go#L889)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary>
> Which drain? The only remaining code below waits for the stopper to stop, which isn't going to happen unless `req.Shutdown` is true. We just weren't ever calling this with `Shutdown: false` before, so nobody noticed.
</details>
ok, so then wouldn't this code more naturally read as

```
if !req.Shutdown {
  return nil
}

s.server.grpc.Stop()
......
```

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/9917)*
<!-- Sent from Reviewable.io -->

tschottdorf commented:
Review status: 0 of 20 files reviewed at latest revision, 3 unresolved discussions, all commit checks successful.

---

*[cli/kv.go, line 237 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/9917#-KTtuBgdy_ANyIoM3G1w:-KTuT8neGMiw813HCja3:bud9anv) ([raw file](https://github.com/cockroachdb/cockroach/blob/053ce1bc478cdd6b373d8fce1e2709ce54d32aa1/cli/kv.go#L237)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> Wait, that's the error you got? Seems dubious since it doesn't contain `increment failed`.
</details>
Oh, sorry. Thought you were wondering about something else. In any case, this code path is tested as I've fixed the examples which exercise it with each change to this message.

```
15:58 $ ./cockroach debug kv inc a b
Error: invalid increment: strconv.ParseInt: parsing "b": invalid syntax
Failed running "debug"
```

---

*[server/admin.go, line 889 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/9917#-KTttOAfXMeYM4UP00-S:-KTuT3D8CJ11GeNtTOld:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/053ce1bc478cdd6b373d8fce1e2709ce54d32aa1/server/admin.go#L889)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> ok, so then wouldn't this code more naturally read as
> 
> ```
> if !req.Shutdown {
>   return nil
> }
> 
> s.server.grpc.Stop()
> ......
> ```
</details>
Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/9917)*
<!-- Sent from Reviewable.io -->

tbg committed:
cli: better errors

Removed `panicGuard` and `mustUsage` and introduced proper error handling,
cleaning up some cruft along the way. However, the primary goal was exposing
better errors to clients. To this end, GRPC errors are now special-cased
and print a legible error. For example,

```
$ ./cockroach node ls # node is not even running
Error: unable to connect or connection lost.

Please check the address and credentials such as certificates (if attempting to
communicate with a secure cluster).

(error code 14: grpc: the connection is unavailable)
Failed running "node"
```

In particular, I removed the `failfast=false` toggle in RPCSender, which (at
least for the cli debug commands, where it is used) would have calls hang
forever, even if the server wasn't even running or the certificates not
specified.

Updated `grpcutil.IsConnectionClosed()` to operate on the `errors.Cause` of
an error.

Reworked `runQuit`:

```
./cockroach quit --host gamma.gce.cockroachdb.com # host alive, but secure
Error: unable to connect or connection lost.

Please check the address and credentials such as certificates (if attempting to
communicate with a secure cluster).

(error code 13: transport is closing)
Failed running "quit"

$ ./cockroach quit # host not running
Error: unable to connect or connection lost.

Please check the address and credentials such as certificates (if attempting to
communicate with a secure cluster).

(error code 14: grpc: the connection is unavailable)
Failed running "quit"

$  $ ./cockroach quit # node running
ok
```

Fixes #9151.
Fixes #9153.

tamird commented:
Reviewed 40 of 40 files at r4.
Review status: all files reviewed at latest revision, 2 unresolved discussions, all commit checks successful.

---

*[cli/kv.go, line 237 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/9917#-KTtuBgdy_ANyIoM3G1w:-KTuoNtIQs_PYazRGzR2:bxbwxn7) ([raw file](https://github.com/cockroachdb/cockroach/blob/053ce1bc478cdd6b373d8fce1e2709ce54d32aa1/cli/kv.go#L237)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary>
> Oh, sorry. Thought you were wondering about something else. In any case, this code path is tested as I've fixed the examples which exercise it with each change to this message.
> 
> ```
> 15:58 $ ./cockroach debug kv inc a b
> Error: invalid increment: strconv.ParseInt: parsing "b": invalid syntax
> Failed running "debug"
> ```
</details>
That is also not the message I'm talking about. Looks like this message is not tested after all.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/9917)*
<!-- Sent from Reviewable.io -->

tschottdorf commented:
Review status: all files reviewed at latest revision, 2 unresolved discussions, all commit checks successful.

---

*[cli/kv.go, line 237 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/9917#-KTtuBgdy_ANyIoM3G1w:-KTwvp4muWopROz_7ta7:b-ehz69k) ([raw file](https://github.com/cockroachdb/cockroach/blob/053ce1bc478cdd6b373d8fce1e2709ce54d32aa1/cli/kv.go#L237)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> That is also not the message I'm talking about. Looks like this message is not tested after all.
</details>
`master`:

```
$ ./cockroach debug kv inc asd
Error: increment failed: roachpb.Batch RPC failed: rpc error: code = 14 desc = grpc: the connection is unavailable
Failed running "debug"
```

Not sure what I copy-pasted above.

https://github.com/cockroachdb/cockroach/pull/9938

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/9917)*
<!-- Sent from Reviewable.io -->
